app/src/main/java/com/github/gotify/service/GotifyMessengerService.kt

```diff
@@ -0,0 +1,176 @@
+            } catch (e: RemoteException) {
+            }
+        }
+
+        private fun unregisterApp(msg: Bundle, clientUid: Int) {
+            val clientPackageName = msg.getString("package").toString()
+            // we only trust unregistered demands from the uid who registered the app
+            if (db.strictIsRegistered(clientPackageName, clientUid)) {
+                Log.i("Unregistering $clientPackageName uid: $clientUid")
+                deleteApp(clientPackageName)
+                db.unregisterApp(clientPackageName, clientUid)
+            }
+        }
+
+        private fun registerApp(msg: Bundle,clientUid: Int) {
+            val clientPackageName = msg.getString("package").toString()
+            if (clientPackageName.isBlank()) {
+                Log.w("Trying to register an app without packageName")
+                return
+            }
+            Log.i("registering $clientPackageName uid: $clientUid")
+            // The app is registered with the same uid : we re-register it
+            // the client may need to create a new app in the server
+            if (db.strictIsRegistered(clientPackageName, clientUid)) {
+                Log.i("$clientPackageName already registered : unregistering to register again")
+                unregisterApp(msg,clientUid)
+            }
+            // The app is registered with a new uid.
+            // User should unregister this app manually
+            // to avoid an app to impersonate another one
+            if (db.isRegistered(clientPackageName)) {
+                Log.w("$clientPackageName already registered with a different uid")
+                return
+            }
+            //
+            val clientService = msg.getString("service").toString()
+            if (clientService.isBlank()) {
+                Log.w("Cannot find the service for $clientPackageName")
+                return
+            }
+            val app = createApp(clientPackageName)
+            if(app == null){
+                Log.w("Cannot create a new app to register")
+                return
+            }
+            Log.i("registering : $clientPackageName $clientUid $clientService ${app.id} ${app.token}")
+            db.registerApp(clientPackageName, clientUid, clientService, app.id, app.token)
+        }
+
+        private fun sendInfo(msg: Message){
+            val clientPackageName = msg.data?.getString("package").toString()
+            Log.i("$clientPackageName is asking for its token and url")
+            // we only trust unregistered demands from the uid who registered the app
+            if (db.strictIsRegistered(clientPackageName, msg.sendingUid)) {
+                // db.getToken also remove the token in the db
+                val token = db.getToken(clientPackageName)
+                try {
+                    val answer = Message.obtain(null, MSG_REGISTER_CLIENT, 0, 0)
+                    answer.data = bundleOf("url" to settings.url(),
```

Comment:
Does the client really need this information? I don't really think so.

Reply:
With FCM/GCM, the client always now the endpoint to send the notification and it is not the case for gotify since it is decentralized.
For example, if I have a mail server and an app and I want to send a notification every time a user has an email, I need to know the endpoint to send the notification. And the endpoint can be registered automatically thanks to this.

Reply:
Are there use-cases for sending messages as the android app themselves? I mean, gotify will just notifiy them with the data they already know.

Reply:
I was thinking of an app sending the endpoint and the token to its server and the server will use the endpoint to notify

Reply:
Okay, that's a good use case (:.